packages/api-plugin-simple-schema/src/queries/introspectSchema.js

```diff
@@ -0,0 +1,75 @@
+import SimpleSchema, { toJsonSchema } from "simpl-schema";
+
+const currentPath = ["$"];
+const inputSchema = new SimpleSchema({
+  shopId: String,
+  schemaName: String
+});
+
+/**
+ * @name updateJsonSchema
+ * @summary Update the JSON Schema to add additional path field for each object
+ * @param {Object} jsonSchema - JSON Schema object
+ * @returns {Object} JSON Schema object with path field added
+ */
+function updateJsonSchema(jsonSchema) {
+  if (jsonSchema.type === "object" && jsonSchema.properties) {
+    const currentObject = jsonSchema.properties;
+    Object.keys(currentObject).forEach((key) => {
+      currentObject[key].path = `${currentPath.join(".")}.${key}`;
+      if (currentObject[key].type === "object") {
+        currentPath.push(key);
+        updateJsonSchema(currentObject[key]); // recursive call
+        currentPath.pop();
+      } else if (currentObject[key].type === "array" && currentObject[key].items && currentObject[key].items.length) {
+        currentPath.push(key);
+        currentPath.push("[0]");
+        const itemObject = currentObject[key].items[0];
+        updateJsonSchema(itemObject); // recursive call
+        itemObject.path = `${currentPath.join(".")}`;
+        currentPath.splice(-2, 2);
+      } else {
+        currentObject[key].path = `${currentPath.join(".")}.${key}`; // when not object/array, we just update the path
+      }
+    });
+  } else if (jsonSchema.type === "array" && jsonSchema.items && jsonSchema.items.length) {
+    currentPath.push("[0]");
+    const itemObject = jsonSchema.items[0];
+    updateJsonSchema(itemObject); // recursive call
+    itemObject.path = `${currentPath.join(".")}`;
+    currentPath.pop();
+  } else {
+    jsonSchema.path = currentPath.join("."); // when not object/array, we just update the path
+  }
+  jsonSchema.path = currentPath.join("."); // top level path
+  return jsonSchema;
+}
+
+/**
+ * @name introspectSchema
+ * @summary Extract the Schema of the required collection from context, convert to Json and return
+ * @param {Object} context - an object containing the per-request state
+ * @param {Object} args - an object of all arguments that were sent by the client
+ * @param {String} [args.shopId] - shopID
+ * @param {String} [args.schemaName] - schemaName to query
+ * @returns {Promise<Object>} JSON Schema object Promise
+ */
+export default async function introspectSchema(context, { shopId, schemaName } = {}) {
+  const validationString = "reaction:legacy:simpleSchema";
+  inputSchema.validate({ shopId, schemaName });
```

Comment:
Since it's impossible to have different schemas for different shops, I would remove the `shopId` input and use global permissions. We use global permissions for "accounts" auth as an example.

Reply:
Copy-pasting the same response from above

As per an earlier discussion with Brent, we decided to give access to some common schemas (like Cart, Product, Order, Account, Promotion) for users in "shop manager" or "owner" groups. 
The migration script and the checkAndSeed.js reflects this plan. As per my understanding, this by default makes the permissions/roles to be added under a specific shopId. Meanwhile the global_roles (userPermission.__global_roles__) are picked from the global groups (groups without a shopId, namely "accounts manager" and "system-manager"). 

So if we remove the shopId param in the introspectSchema query, then we have to re-visit the earlier plan. At least we have to include these roles (selected schema access roles) in the global-groups in migration/checkAndSeed code.

Please let me know if I should go ahead with those changes as well.

Reply:
same update as above comment